File: 2. LeetCode/7. Reverse Integer.cpp
#include <climits> // INT_MAX, INT_MIN
#include <iostream>

class Solution {
public:
    int reverse(int x) {
        int result = 0;

        while (x != 0) {
            int pop = x % 10;
            x /= 10;

            // 오버플로우 체크
            if (result > INT_MAX / 10 || (result == INT_MAX / 10 && pop > 7))
                return 0;
            if (result < INT_MIN / 10 || (result == INT_MIN / 10 && pop < -8))
                return 0;

            result = result * 10 + pop;
        }

        return result;
    }
};

int main() {
    Solution solution;
    int testValue = 123; // 테스트할 값
    int reversedValue = solution.reverse(testValue);

    std::cout << "Original: " << testValue << std::endl;
    std::cout << "Reversed: " << reversedValue << std::endl;

    return 0;
}

File: 2. LeetCode/7. Reverse Integer.py
class Solution:
    def reverse(self, x: int) -> int:
        # 오류 범위 정의
        INT_MAX = 2**31 - 1
        INT_MIN = -(2**31)

        # 음수인지 확인
        sign = -1 if x < 0 else 1
        x = abs(x)

        # 숫자 뒤집고 부호 적용
        reversed_number = int(str(x)[::-1]) * sign

        # 오버플로우 검사
        if reversed_number < INT_MIN or reversed_number > INT_MAX:
            return 0

        return reversed_number


if __name__ == "__main__":
    sol = Solution()
    print(sol.reverse(123))  # 321
